src/test/java/com/melon/portfoliomanager/UserControllerTest.java

```diff
@@ -0,0 +1,81 @@
+package com.melon.portfoliomanager;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.melon.portfoliomanager.controllers.UserController;
+import com.melon.portfoliomanager.exceptions.NullFieldException;
+import com.melon.portfoliomanager.exceptions.UsernameAlreadyUsedException;
+import com.melon.portfoliomanager.models.User;
+import com.melon.portfoliomanager.repositories.UserRepository;
+import com.melon.portfoliomanager.services.UserService;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+
+import java.util.ArrayList;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+public class UserControllerTest {
+
+    @InjectMocks
+    UserController userController;
+
+    @Mock
+    UserService userService;
+
+    ObjectMapper objectMapper = new ObjectMapper();
+
+    MockMvc mockMvc;
+
+    @BeforeEach
+    public void setUp() throws Exception {
+        MockitoAnnotations.initMocks(this);
+        mockMvc = MockMvcBuilders.standaloneSetup(userController).build();
```

Comment:
Hm, we can really use the ```MockMvc``` but it is not as close to the real scenario as the ```WebTestClient``` is. Also we can use the ```WebClient``` class for sending http queries to external services (another microservices, APIs, etc.) - this is something that we need to discuss but my opinion is that we should stick up to the ```WebTestClient``` for the test classes and ```WebClient``` for the production code.
(NB: ```WebClient``` is the modern alternative to the ```RestTemplate``` class which is already deprecated)